File: support.cpp
#include "support.h"
#include "mathmore.h"
#include <cmath>
#include <string>
#include <iostream>
#include <fstream>
#include <algorithm>




/* ----------------------POLYNOMIAL ESTIMATOR NO SPLITTING-------------------------------- */
double Support::estimate() const
{
    if ( pmin < Ratio / n ) // const
    {
        double result = 0.0;
        for ( const auto & pair : fin )
                result += getCoeff(pair.first) * pair.second;
        return result;
    }
    else
        return estimate_plug();
}

double Support::getCoeff( int N ) const
{
    if ( N == 0 ) return 0;
    if ( N > L ) return 1;

    double A = ( Ratio + n*pmin ) / ( Ratio - n*pmin );
    ChebMore cheb(L, 1, -A); // polynomial of cos L arccos(t-A)
    std::vector<double> a = cheb.expand(); // Expand: cos L arccos(x-A)=sum_i a[i]*x^i
    double amp = cheb.evaluate(0); // cos L arccos(-A)

    double result = -a[N]/amp;
    double s = 2 / ( Ratio - n * pmin );
    for ( int i = 1; i <= N; ++i )
        result *= i * s;

    return (1+result);
}
/* ----------------------END POLYNOMIAL ESTIMATOR NO SPLITTING-------------------------------- */





/* ----------------------OTHER ESTIMATORS-------------------------------- */
// Ref for TG, CL1, CL2: "Estimating the Number of Classes via Sample Coverage, Chao-Lee, 1992"
// Ref for JK: "Robust Estimation of Population Size When Capture Probabilities Vary Among Animals. K. P. Burnham and W. S. Overton, 1979"

double Support::estimate_plug() const
{
    double result = 0.0;
    for ( const auto & pair : fin )
        result += pair.second;
    return result;
}

double Support::coverage_TG() const
{
    double n1 = 0;
    for ( const auto & pair : fin )
        if ( pair.first == 1 )
        {
            n1 = pair.second;
            break;
        }
    return (1 - n1/n);
}

double Support::estimate_TG() const
{
    return ( estimate_plug() / coverage_TG() );
}


double Support::estimate_J1() const
{
    double f1 = 0;
    for ( const auto & pair : fin )
        if ( pair.first == 1 )
        {
            f1 = pair.second;
            break;
        }
    return ( estimate_plug() + (n-1.0)/n*f1 );
}


double Support::estimate_Chao1() const 
{
    double f1 = 0;
    double f2 = 0;
    for ( const auto & pair : fin )
        if ( pair.first == 1 )
        {
            f1 = pair.second;
            break;
        }
    for ( const auto & pair : fin )
        if ( pair.first == 2 )
        {
            f2 = pair.second;
            break;
        }
    return estimate_plug()+f1*(f1-1)/(2*(f2+1));
}

double Support::estimate_CL1() const // for small coefficient of variation
{
    double N1 = estimate_TG();
    double C = coverage_TG();

    double sum = 0.0;
    for ( const auto & pair : fin )
        sum += pair.first*( pair.first-1 )*pair.second;
    double gamma_sq = std::max( estimate_TG()*sum/n/(n-1) - 1 , 0.0 );
    return N1 + n*(1-C)/C*gamma_sq;
}


double Support::estimate_CL2() const // for large coefficient of variation
{
    double N1 = estimate_TG();
    double C = coverage_TG();
    
    double sum = 0.0;
    for ( const auto & pair : fin )
        sum += pair.first*( pair.first-1 )*pair.second;
    double gamma_sq = std::max( estimate_TG()*sum/n/(n-1) - 1 , 0.0 );
    double gamma_sq2 = std::max( gamma_sq*( 1 + n*(1-C)*sum/n/(n-1)/C ), 0.0 );
	
    return N1 + n*(1-C)/C*gamma_sq2;
}

/* ----------------------END OTHER ESTIMATORS-------------------------------- */





Support::Support()
{
}


Support::Support(double p_min)
    :Support()
{
    pmin = p_min;
    setDegree( 0.45*log(1.0/pmin) );      // Default L = 0.45 * log(1/pmin)
    setInterval( 0.5*log(1.0/pmin) );     // Default interval [pmin, 0.5*log(1/pmin)/n]
}


void Support::setFin(const std::string filename) 
{
    fin.clear();
    n = 0;
    std::ifstream infile;
    infile.open( filename.c_str() );
    int freq, cnt;
    while ( (infile>>freq).good() )
    {
        infile >> cnt;
        fin.push_back( std::make_pair( freq, cnt ) );
        n += (freq * cnt);
    }
    infile.close();
}

void Support::setFin(const std::vector<int> &freq_in, const std::vector<int> &cnt_in)
{
    fin.clear();
    n = 0;
    for ( unsigned i = 0; i < freq_in.size(); i++)
    {
        int freq = freq_in[i], cnt = cnt_in[i];
        fin.push_back( std::make_pair( freq, cnt ) );
        n += (freq * cnt);
    }
}


void Support::setHist( const std::vector<unsigned> &hist )
{
    fin.clear();
    n = 0;
    
    std::map<int, int> fin_map;
    for ( const auto & freq : hist )
    {
        auto iter = fin_map.find( freq );
        if ( iter == fin_map.end() )
            fin_map.insert( std::make_pair( freq,1 ) );
        else
            ++(iter->second);
    }
    
    for ( auto it = fin_map.begin(); it != fin_map.end(); ++it )
    {
        int freq = it->first, cnt = it->second;
        fin.push_back( std::make_pair( freq, cnt ) );
        n += (freq * cnt);
    }
}

void Support::setHist(const std::string filename) 
{
    fin.clear();
    n = 0;
    std::map<int, int> fin_map;
    
    std::ifstream infile;
    infile.open( filename.c_str() );
    int freq;
    while ( (infile>>freq).good() )
    {
        if ( freq == 0)
            continue;

        auto iter = fin_map.find( freq );
        if ( iter == fin_map.end() )
            fin_map.insert( std::make_pair( freq,1 ) );
        else
            ++(iter->second);
    }
    infile.close();

    for ( auto it = fin_map.begin(); it != fin_map.end(); ++it )
    {
        int freq = it->first, cnt = it->second;
        fin.push_back( std::make_pair( freq, cnt ) );
        n += (freq * cnt);
    }
}

File: support.py
#!/usr/bin/env python3
"""
Main libraray for support size estimation
"""

from math import log, floor
import numpy as np
import scipy.special



class Support():
    """
    Support estimator
    """
    def __init__(self, pmin, L=None, M=None):
        """
        Args:
        pmin: float, required
        =1/k. Preset minimum non-zero mass

        L: int
        Polynoimal degree. Default c0*log(k)

        M: float
        M/n is the right-end of approximation interval. Default c1*log(k)

        """
        self.pmin = pmin
        self.degree = L if L != None else floor(0.45*log(1./pmin))
        self.ratio = M if M != None else 0.5*log(1./pmin)

    def estimate(self, fin):
        """
        Our rate-optimal estimator from a given fingerprint

        Args:
        fin: list of tuples (frequency, count)
        fingerprint of samples, fin[i] is the number of symbols that appeared exactly i times

        Return:
        the estimated entropy (bits)
        """
        # get total sample size
        num = get_sample_size(fin)

        if self.pmin < self.ratio / num:
            # get linear estimator coefficients
            cheb = np.polynomial.chebyshev.Chebyshev.basis(self.degree)
            cheb_coeffs = np.polynomial.chebyshev.cheb2poly(cheb.coef)
            shift = (self.ratio + num*self.pmin) / (self.ratio - num*self.pmin)
            a_coeffs = shift_polynomial(cheb_coeffs, -shift)
            g_coeffs = -a_coeffs / a_coeffs[0]
            g_coeffs[0] = 0
            scaling = 2. / (self.ratio-num*self.pmin)
            for j in range(1, self.degree+1):
                for i in range(1, j+1):
                    g_coeffs[j] *= (i*scaling)
                g_coeffs[j] += 1

            # get estimate
            s_estimate = 0
            for freq, cnt in fin:
                if freq > self.degree: # plug-in
                    s_estimate += 1*cnt
                else: # polynomial
                    s_estimate += g_coeffs[freq]*cnt
        else:
            s_estimate = self.estimate_plug(fin)

        return s_estimate

    def estimate_plug(self, fin):
        """
        Plug-in estimate of support size: number of seen symbols
        """
        s_estimate = 0
        for freq, cnt in fin:
            s_estimate += cnt
        return s_estimate

    def coverage_turing_good(self, fin):
        """
        Turing-Good coverage
        """
        num = get_sample_size(fin)
        fin1 = 0
        for freq, cnt in fin:
            if freq == 1:
                fin1 = cnt
                break
        return 1.-1.*fin1/num

    def estimate_turing_good(self, fin):
        """
        Turing-Good estimator
        Ref: "The Population Frequencies of Species and the Estimation of Population Parameters",
        I. J. Good, 1953
        """
        return 1.*self.estimate_plug(fin)/self.coverage_turing_good(fin)

    def estimate_jackknife1(self, fin):
        """
        First order Jackknife
        Ref: "Robust Estimation of Population Size When Capture Probabilities Vary Among Animals",
        K. P. Burnham and W. S. Overton, 1979
        """
        num = get_sample_size(fin)
        fin1 = 0
        for freq, cnt in fin:
            if freq == 1:
                fin1 = cnt
                break
        return self.estimate_plug(fin) + (num-1.0)/num*fin1
        
    def estimate_chao1(self, fin):
        """
        Chao 1 estimator
        Ref original Chao 1: "Nonparametric estimation of the number of classes in a population."
        A. Chao, 1984
        Ref this bias-corrected Chao 1: "Estimating species richness",
        N. J. Gotelli and R. K. Colwell, 2011
        """
        fin1 = 0
        fin2 = 0
        for freq, cnt in fin:
            if freq == 1:
                fin1 = cnt
                break
        for freq, cnt in fin:
            if freq == 2:
                fin2 = cnt
                break
        return self.estimate_plug(fin)+fin1*(fin1-1)/(2*(fin2+1))

    def estimate_chao_lee1(self, fin):
        """
        Chao-Lee 1, for small coefficient of variation
        Ref: "Estimating the Number of Classes via Sample Coverage", A. Chao and S. Lee, 1992
        """
        num = get_sample_size(fin)
        estimate1 = self.estimate_turing_good(fin)
        coverage = self.coverage_turing_good(fin)

        sums = 0.
        for freq, cnt in fin:
            sums += freq*(freq-1)*cnt
        gamma_sq = max(1.*estimate1*sums/num/(num-1)-1, 0)
        return estimate1 + num*(1-coverage)/coverage*gamma_sq

    def estimate_chao_lee2(self, fin):
        """
        Chao-Lee 2, for large coefficient of variation
        Ref: "Estimating the Number of Classes via Sample Coverage", A. Chao and S. Lee, 1992
        """
        num = get_sample_size(fin)
        estimate1 = self.estimate_turing_good(fin)
        coverage = self.coverage_turing_good(fin)

        sums = 0.
        for freq, cnt in fin:
            sums += freq*(freq-1)*cnt
        gamma_sq = max(1.*estimate1*sums/num/(num-1)-1, 0)
        gamma_sq2 = max(gamma_sq*(1+num*(1-coverage)*sums/num/(num-1)/coverage), 0)
        return estimate1 + num*(1-coverage)/coverage*gamma_sq2

def get_sample_size(fin):
    """
    get total sample size from a given fingerprint
    """
    num = 0
    for freq, cnt in fin:
        num += freq * cnt
    return num


def sample_to_fin(samples):
    """
    Return a fingerprint from samples
    """
    return hist_to_fin(sample_to_hist(samples))


def hist_to_fin(hist):
    """
    Return a fingerprint from histogram
    """
    fin = Counter()
    for freq in hist:
        fin[freq] += 1
    return fin.items()

def sample_to_hist(samples):
    """
    Return a histogram of samples
    """
    freq = Counter()
    for symbol in samples:
        freq[symbol] += 1
    return np.asarray(list(freq.values()))


def shift_polynomial(coeffs, shift):
    """
    Return coefficients of a shifted polynomial
    P(x) = p(x+shift)

    Args:
    coeffs: list of float
    coefficients of p

    shift: float
    right shift of polynomial

    Returns:
    coefficients of P
    """
    length = len(coeffs)
    coeffs_shift = np.zeros(length)
    for i in range(length):
        for j in range(i+1):
            coeffs_shift[j] += coeffs[i] * (shift**(i-j)) * scipy.special.binom(i, j)
    return coeffs_shift


class Counter(dict):
    """
    Class for counting items
    """
    def __missing__(self, key):
        return 0

